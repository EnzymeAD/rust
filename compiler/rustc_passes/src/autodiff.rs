//! Detect autodiff blocks
//!
//! External blocks can be decorated with an autodiff attribute, generating functions
//! during compilation with automatic differentiation from a source function.
//!
//! # Example
//!
//! ```rust
//! fn rosenbrock(a: f32, b: f32, x: f32, y: f32) -> f32 {
//!     let (z, w) = (a-x, y-x*x);
//!
//!     z*z + b*w*w
//! }
//!
//! #[autodiff(rosenbrock, mode = "forward")]
//! extern "C" {
//!     fn dx_rosenbrock(a: f32, b: f32, x: f32, y: f32, d_x: &mut f32);
//!     fn dy_rosenbrock(a: f32, b: f32, x: f32, y: f32, d_y: &mut f32);
//! }
//!

use rustc_hir::def_id::LocalDefId;
use rustc_hir::itemlikevisit::ItemLikeVisitor;
use rustc_hir::{ForeignItem, ForeignItemKind, HirId, ImplItem, Item, ItemKind, TraitItem};
use rustc_middle::ty::query::Providers;
use rustc_middle::ty::TyCtxt;
use rustc_ast::{Lit, LitKind, MetaItem, MetaItemKind, NestedMetaItem, Path};
use rustc_data_structures::fx::FxHashMap;
use rustc_middle::metadata::{DiffItems, DiffMode, DiffActivity, DiffItem};
use rustc_span::{symbol::{sym, Ident}, Span};
use rustc_target::spec::abi::Abi;

#[derive(Clone)]
struct AdTask {
    mode: DiffMode,
    ret: DiffActivity,
    inputs: Vec<DiffActivity>,
}

#[allow(dead_code)]
struct AutodiffContext<'tcx> {
    tcx: TyCtxt<'tcx>,
    methods: Vec<((Path, LocalDefId), AdTask, &'tcx ForeignItem<'tcx>, Span)>,
    source_candidates: FxHashMap<(Ident, LocalDefId), HirId>,
}

impl<'tcx> ItemLikeVisitor<'tcx> for AutodiffContext<'tcx> {
    fn visit_item(&mut self, item: &'tcx Item<'tcx>) {
        if let ItemKind::Fn(sig, _, _) = &item.kind {
            if !matches!(sig.header.abi, Abi::C { .. }) {
                self.source_candidates.insert(
                    (item.ident, self.tcx.hir().get_parent_item(item.hir_id())),
                    item.hir_id(),
                    );
            }

            return;
        }

        // skip all items except extern blocks
        let ItemKind::ForeignMod { items, .. } = item.kind else {
            return
        };

        let id = item.hir_id();
        let attrs = self.tcx.hir().attrs(id);
        let attrs = attrs
            .into_iter()
            .filter(|attr| attr.name_or_empty() == sym::autodiff)
            .collect::<Vec<_>>();

        // check for exactly one autodiff attribute on extern block
        let attr = match &attrs[..] {
            &[] => return,
            &[elm] => elm,
            _ => {
                self.tcx
                    .sess
                    .struct_span_err(item.span, "autodiff attribute can only be applied once")
                    .span_label(item.span, "more than one")
                    .emit();

                return;
            }
        };

        let list = attr.meta_item_list().unwrap_or_default();
        //assert!(list.len() > 3);
        let tmp_ret  = &list[2];
        //let tmp_args = &list[3..];

        dbg!(tmp_ret);

        let ret_symbol = match tmp_ret {
            NestedMetaItem::MetaItem(MetaItem {
                path: ref p2,
                kind: MetaItemKind::Word,
                ..
            }) => {
                let id = p2.segments.first().unwrap().ident;
                dbg!(&id);
                id
            },
            _ => {
                self.tcx
                    .sess
                    .struct_span_err(
                        item.span,
                        "autodiff attribute must contain the return activity",
                        )
                    .span_label(item.span, "missing return activity")
                    .emit();

                return;
            },
        };


        let (path, mode) = match &list[0..=1] {
            &[
                NestedMetaItem::MetaItem(MetaItem { ref path, kind: MetaItemKind::Word, .. }),
                NestedMetaItem::MetaItem(MetaItem {
                    path: ref p2,
                    kind: MetaItemKind::NameValue(Lit { kind: LitKind::Str(mode, _), .. }),
                    ..
                }),
            ] if *p2 == sym::mode => (path, mode),
            _ => {
                self.tcx
                    .sess
                    .struct_span_err(
                        item.span,
                        "autodiff attribute must contain the source function",
                        )
                    .span_label(item.span, "empty argument list")
                    .emit();

                return;
            }
        };

        let ret_activity = match ret_symbol.as_str() {
            "None" => DiffActivity::None,
            "Active" => DiffActivity::Active,
            "Const" => DiffActivity::Const,
            "Duplicated" => DiffActivity::Duplicated,
            "DuplicatedNoNeed" => DiffActivity::DuplicatedNoNeed,
            _ => {
                self.tcx
                    .sess
                    .struct_span_err(item.span, "unknown return activity")
                    .span_label(item.span, "invalid return activity")
                    .emit();
                return;
            }
        };

        let mut input_activity: Vec<DiffActivity> = vec![];
        for foo in &list[3..] {
            let foo_symbol = match foo {
                NestedMetaItem::MetaItem(MetaItem {
                    path: ref p2,
                    kind: MetaItemKind::Word,
                    ..
                }) => {
                    let id = p2.segments.first().unwrap().ident;
                    dbg!(&id);
                    id
                },
                _ => {
                    self.tcx
                        .sess
                        .struct_span_err(
                            item.span,
                            "autodiff attribute must contain the return activity",
                            )
                        .span_label(item.span, "missing return activity")
                        .emit();

                    return;
                },
            };
            let act = match foo_symbol.as_str() {
                "None" => DiffActivity::None,
                "Active" => DiffActivity::Active,
                "Const" => DiffActivity::Const,
                "Duplicated" => DiffActivity::Duplicated,
                "DuplicatedNoNeed" => DiffActivity::DuplicatedNoNeed,
                _ => {
                    self.tcx
                        .sess
                        .struct_span_err(item.span, "unknown return activity")
                        .span_label(item.span, "invalid input activity")
                        .emit();
                    return;
                }
            };
            input_activity.push(act);
        }

        // parse mode
        let mode = match mode.as_str() {//map(|x| x.as_str()) {
            "forward" => DiffMode::Forward,
            "reverse" => DiffMode::Reverse,
            _ => {
                self.tcx
                    .sess
                    .struct_span_err(item.span, "mode should be either forward or reverse")
                    .span_label(item.span, "invalid mode")
                    .emit();

                return;
            }
        };

        if mode == DiffMode::Forward {
            if ret_activity == DiffActivity::Active {
                self.tcx
                    .sess
                    .struct_span_err(item.span, "Forward Mode is incompatible with Active ret")
                    .span_label(item.span, "invalid return activity")
                    .emit();
                return;
            }
            if input_activity.iter().filter(|&x| *x == DiffActivity::Active).count() > 0 {
                self.tcx
                    .sess
                    .struct_span_err(item.span, "Forward Mode is incompatible with Active args")
                    .span_label(item.span, "invalid input activity")
                    .emit();
                return;
            }
        }

        if mode == DiffMode::Reverse {
            if ret_activity == DiffActivity::Duplicated || ret_activity == DiffActivity::DuplicatedNoNeed {
                self.tcx
                    .sess
                    .struct_span_err(item.span, "Reverse Mode is only compatible with Active, None, or Const ret")
                    .span_label(item.span, "invalid return activity")
                    .emit();
                return;
            }
        }

        let task = AdTask {
            mode,
            ret: ret_activity,
            inputs: input_activity,
        };


        let parent_id = self.tcx.hir().get_parent_item(HirId::make_owner(item.def_id));
        self.methods.extend(
            items.into_iter()
            .map(|x| self.tcx.hir().foreign_item(x.id))
            .map(|x| ((path.clone(), parent_id), task.clone(), x, item.span))
            );

        //dbg!(&parent_id);
        //self.tcx.hir().visit_item_likes_in_module(parent_id, self);

        //visit_path(

        //visit_path
        //maybe_resolve_path
        }

        fn visit_trait_item(&mut self, _trait_item: &'tcx TraitItem<'tcx>) {
            // Entry fn is never a trait item.
        }

        fn visit_impl_item(&mut self, _impl_item: &'tcx ImplItem<'tcx>) {
            // Entry fn is never a trait item.
        }

        fn visit_foreign_item(&mut self, _item: &'tcx ForeignItem<'tcx>) {}
    }

#[allow(dead_code)]
    fn get_autodiff_functions(tcx: TyCtxt<'_>, (): ()) -> DiffItems {
        let mut ctxt =
            AutodiffContext { tcx, methods: Vec::new(), source_candidates: FxHashMap::default() };

        tcx.hir().visit_all_item_likes(&mut ctxt);

        let mut elms = Vec::new();
        for method in ctxt.methods {
            // get source function's def_id
            let ident = method.0.0.segments.last().unwrap().ident;
            let Some(source) = ctxt.source_candidates.get(&(ident, method.0.1)) else {
                tcx
                    .sess
                    .struct_span_err(
                        method.3,
                        "method not found for autodiff attribute",
                        )
                    .span_label(method.0.0.span, "unknown method")
                    .emit();

                return DiffItems::default();
            };

            //let source = tcx.hir().fn_decl_by_hir_id(*source).unwrap();

            let def_id = tcx.hir().local_def_id(*source).to_def_id();
            let source_args = tcx.fn_arg_names(def_id);

            let _fn_args = match method.2.kind {
                ForeignItemKind::Fn(_, args, _) => args,
                _ => unreachable!(),
            };

            // skip beginning of function parameters, make sure they are the same
            //  if &fn_args[..source_args.len()] != source_args {
            //      tcx
            //          .sess
            //          .struct_span_err(
            //              method.0.0.span,
            //              "method has to begin with same arguments",
            //              )
            //          .span_label(method.0.0.span, "arguments differ")
            //          .emit();

            //      return DiffItems::default();
            //  }
            assert!(method.1.inputs.len() == source_args.len());

            elms.push(DiffItem {
                source: def_id,
                target: method.2.ident.to_string(),
                mode: method.1.mode,
                ret_activity: method.1.ret,
                input_activity: method.1.inputs,
            });
        }

        elms
    }

    pub fn provide(providers: &mut Providers) {
        providers.autodiff_functions = get_autodiff_functions;
    }
