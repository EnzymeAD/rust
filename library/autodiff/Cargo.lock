# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "autocfg"
version = "1.1.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d468802bab17cbc0cc575e9b053f41e72aa36bfa6b7f55e3529ffa43161b97fa"

[[package]]
name = "autodiff"
version = "0.1.0"
dependencies = [
 "macrotest",
 "ndarray",
 "proc-macro-error",
 "proc-macro2",
 "quote",
 "syn 1.0.109",
 "trybuild",
]

[[package]]
name = "basic-toml"
version = "0.1.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "24c12265665aebaa236af9bbe266681bcc9c5666192119e3d8335cf083aca26f"
dependencies = [
 "serde",
]

[[package]]
name = "diff"
version = "0.1.13"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "56254986775e3233ffa9c4d7d3faaf6d36a2c09d30b20687e9f88bc8bafc16c8"

[[package]]
name = "glob"
version = "0.3.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d2fabcfbdc87f4758337ca535fb41a6d701b65693ce38287d856d1674551ec9b"

[[package]]
name = "itoa"
version = "1.0.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "af150ab688ff2122fcef229be89cb50dd66af9e01a4ff320cc137eecc9bacc38"

[[package]]
name = "macrotest"
version = "1.0.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "7489ae0986ce45414b7b3122c2e316661343ecf396b206e3e15f07c846616f10"
dependencies = [
 "diff",
 "glob",
 "prettyplease",
 "serde",
 "serde_json",
 "syn 1.0.109",
 "toml",
]

[[package]]
name = "matrixmultiply"
version = "0.3.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "7574c1cf36da4798ab73da5b215bbf444f50718207754cb522201d78d1cd0ff2"
dependencies = [
 "autocfg",
 "rawpointer",
]

[[package]]
name = "ndarray"
version = "0.15.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "adb12d4e967ec485a5f71c6311fe28158e9d6f4bc4a447b474184d0f91a8fa32"
dependencies = [
 "matrixmultiply",
 "num-complex",
 "num-integer",
 "num-traits",
 "rawpointer",
]

[[package]]
name = "num-complex"
version = "0.4.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1ba157ca0885411de85d6ca030ba7e2a83a28636056c7c699b07c8b6f7383214"
dependencies = [
 "num-traits",
]

[[package]]
name = "num-integer"
version = "0.1.45"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "225d3389fb3509a24c93f5c29eb6bde2586b98d9f016636dff58d7c6f7569cd9"
dependencies = [
 "autocfg",
 "num-traits",
]

[[package]]
name = "num-traits"
version = "0.2.17"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "39e3200413f237f41ab11ad6d161bc7239c84dcb631773ccd7de3dfe4b5c267c"
dependencies = [
 "autocfg",
]

[[package]]
name = "once_cell"
version = "1.18.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "dd8b5dd2ae5ed71462c540258bedcb51965123ad7e7ccf4b9a8cafaa4a63576d"

[[package]]
name = "prettyplease"
version = "0.1.25"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6c8646e95016a7a6c4adea95bafa8a16baab64b583356217f2c85db4a39d9a86"
dependencies = [
 "proc-macro2",
 "syn 1.0.109",
]

[[package]]
name = "proc-macro-error"
version = "1.0.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "da25490ff9892aab3fcf7c36f08cfb902dd3e71ca0f9f9517bea02a73a5ce38c"
dependencies = [
 "proc-macro-error-attr",
 "proc-macro2",
 "quote",
 "syn 1.0.109",
 "version_check",
]

[[package]]
name = "proc-macro-error-attr"
version = "1.0.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a1be40180e52ecc98ad80b184934baf3d0d29f979574e439af5a55274b35f869"
dependencies = [
 "proc-macro2",
 "quote",
 "version_check",
]

[[package]]
name = "proc-macro2"
version = "1.0.69"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "134c189feb4956b20f6f547d2cf727d4c0fe06722b20a0eec87ed445a97f92da"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.33"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5267fca4496028628a95160fc423a33e8b2e6af8a5302579e322e4b520293cae"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "rawpointer"
version = "0.2.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "60a357793950651c4ed0f3f52338f53b2f809f32d83a07f72909fa13e4c6c1e3"

[[package]]
name = "ryu"
version = "1.0.15"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1ad4cc8da4ef723ed60bced201181d83791ad433213d8c24efffda1eec85d741"

[[package]]
name = "serde"
version = "1.0.190"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "91d3c334ca1ee894a2c6f6ad698fe8c435b76d504b13d436f0685d648d6d96f7"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.190"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "67c5609f394e5c2bd7fc51efda478004ea80ef42fee983d5c67a65e34f32c0e3"
dependencies = [
 "proc-macro2",
 "quote",
 "syn 2.0.38",
]

[[package]]
name = "serde_json"
version = "1.0.107"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6b420ce6e3d8bd882e9b243c6eed35dbc9a6110c9769e74b584e0d68d1f20c65"
dependencies = [
 "itoa",
 "ryu",
 "serde",
]

[[package]]
name = "syn"
version = "1.0.109"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "72b64191b275b66ffe2469e8af2c1cfe3bafa67b529ead792a6d0160888b4237"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "syn"
version = "2.0.38"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e96b79aaa137db8f61e26363a0c9b47d8b4ec75da28b7d1d614c2303e232408b"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "termcolor"
version = "1.3.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6093bad37da69aab9d123a8091e4be0aa4a03e4d601ec641c327398315f62b64"
dependencies = [
 "winapi-util",
]

[[package]]
name = "toml"
version = "0.5.11"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f4f7f0dd8d50a853a531c426359045b1998f04219d88799810762cd4ad314234"
dependencies = [
 "serde",
]

[[package]]
name = "trybuild"
version = "1.0.85"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "196a58260a906cedb9bf6d8034b6379d0c11f552416960452f267402ceeddff1"
dependencies = [
 "basic-toml",
 "glob",
 "once_cell",
 "serde",
 "serde_derive",
 "serde_json",
 "termcolor",
]

[[package]]
name = "unicode-ident"
version = "1.0.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3354b9ac3fae1ff6755cb6db53683adb661634f67557942dea4facebec0fee4b"

[[package]]
name = "version_check"
version = "0.9.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "49874b5167b65d7193b8aba1567f5c7d93d001cafc34600cee003eda787e483f"

[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"

[[package]]
name = "winapi-util"
version = "0.1.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f29e6f9198ba0d26b4c9f07dbe6f9ed633e1f3d5b8b414090084349e46a52596"
dependencies = [
 "winapi",
]

[[package]]
name = "winapi-x86_64-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"
