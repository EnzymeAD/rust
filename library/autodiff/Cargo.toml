[package]
name = "autodiff"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html


[profile.release]
lto = "fat"

[profile.dev]
lto = "fat"

[lib]
name = "autodiff"
proc-macro = true

[dependencies]
quote = "1.0"
proc-macro2 = "1"
proc-macro-error = "1"
syn = { version = "1", features = ["extra-traits", "full", "visit", "visit-mut"]}

[dev-dependencies]
macrotest = "1"
trybuild = "1"
ndarray = "0.15"
trycmd = "0.14"
