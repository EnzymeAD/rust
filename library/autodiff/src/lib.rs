use proc_macro::TokenStream;
use proc_macro_error::proc_macro_error;
use quote::quote;

mod gen;
mod parser;

#[proc_macro_attribute]
#[proc_macro_error]
pub fn autodiff(args: TokenStream, input: TokenStream) -> TokenStream {
    let mut params = parser::parse(args.into(), input.clone().into());
    let (primal, adjoint) = (gen::primal_fnc(&mut params), gen::adjoint_fnc(&params));

    let res = quote!(
        #primal
        #adjoint
    );

    res.into()
}

#[test]
pub fn expanding() {
    macrotest::expand("tests/expand/*.rs");
}

#[test]
fn ui() {
    let t = trybuild::TestCases::new();
    t.compile_fail("tests/ui/*.rs");
}
