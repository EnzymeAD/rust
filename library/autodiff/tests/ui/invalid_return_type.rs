use autodiff::autodiff;

#[autodiff(d_sin, Reverse, Active)]
fn active_but_no_return(#[active] x: f32) {
}

#[autodiff(d_sin, Reverse, Active)]
fn invalid_primal_value(#[active] x: f32, #[active] y: Vec<f32>, #[active] z: Tensor, y_tang: f32) -> (i32, f32);

#[autodiff(d_sin, Forward, Duplicated)]
fn invalid_forward_return(#[dup] x: &f32, tx: &f32, #[dup] y: &Vec<f32>, ty: &Vec<f32>, #[dup] z: &Tensor, tz: &Tensor) -> (f32, f32, f32);

#[autodiff(d_sin, Forward, DuplicatedNoNeed)]
fn invalid_forward_return(#[dup] x: &f32, tx: &f32, #[dup] y: &Vec<f32>, ty: &Vec<f32>, #[dup] z: &Tensor, tz: &Tensor) -> (f32, f32);

fn main() {}
