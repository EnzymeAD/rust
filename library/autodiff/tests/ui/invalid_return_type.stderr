error: no return type
 --> tests/ui/invalid_return_type.rs:4:1
  |
4 | fn active_but_no_return(#[active] x: f32) {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: `#[autodiff]` non-const return activity but no return type

error: invalid output
 --> tests/ui/invalid_return_type.rs:8:100
  |
8 | fn invalid_primal_value(#[active] x: f32, #[active] y: Vec<f32>, #[active] z: Tensor, y_tang: f32) -> (i32, f32);
  |                                                                                                    ^^^^^^^^^^^^^
  |
  = help: `#[autodiff]` expected (f32, Vec < f32 >, Tensor,)

error: invalid output
  --> tests/ui/invalid_return_type.rs:11:121
   |
11 | fn invalid_forward_return(#[dup] x: &f32, tx: &f32, #[dup] y: &Vec<f32>, ty: &Vec<f32>, #[dup] z: &Tensor, tz: &Tensor) -> (f32, f32, f32);
   |                                                                                                                         ^^^^^^^^^^^^^^^^^^
   |
   = help: `#[autodiff]` expected (f32, f32, f32, f32,)

error: invalid output
  --> tests/ui/invalid_return_type.rs:14:121
   |
14 | fn invalid_forward_return(#[dup] x: &f32, tx: &f32, #[dup] y: &Vec<f32>, ty: &Vec<f32>, #[dup] z: &Tensor, tz: &Tensor) -> (f32, f32);
   |                                                                                                                         ^^^^^^^^^^^^^
   |
   = help: `#[autodiff]` expected (f32, f32, f32,)
